backend/gene_info/api/ncbi_provider.py

```diff
@@ -44,38 +44,66 @@ def fetch_gene_info_tree(self, uid):
         """
 
         if self.api_key is None:
             raise NCBIAPIException
-        search_url = f"{self.base_ncbi_uri}esearch.fcgi?db=gene&term={geneID}{self.api_key}&retmode=json"
+        # first search with gene ENSEMBL id
         try:
-            search_response = urllib.request.urlopen(search_url).read()
-        except Exception:
+            search_response = self._search_gene_uid(geneID)
+        except NCBIUnexpectedResultException:
             raise NCBIUnexpectedResultException
-        search_result = self._load_search_result(search_response)
-        if not search_result:
+
+        # search with gene name if needed
+        if not self._is_valid_search_result(search_response) and gene != "":
+            try:
+                search_response = self._search_gene_uid(gene)
+                if self._is_valid_search_result(search_response):
+                    return (int(search_response["esearchresult"]["idlist"][0]), True)
+                else:
+                    logging.error(f"Unexpected NCBI search result, got {search_response}")
+                    raise NCBIUnexpectedResultException
+            except NCBIUnexpectedResultException:
+                raise NCBIUnexpectedResultException
+        elif not self._is_valid_search_result(search_response):
+            logging.error(f"Unexpected NCBI search result, got {search_response}")
             raise NCBIUnexpectedResultException
         else:
-            return search_result
+            return (int(search_response["esearchresult"]["idlist"][0]), False)
 
-    def _load_search_result(self, response):
-        search_result = json.loads(response)
+    def _search_gene_uid(self, term):
+        """Conducts an Esearch using NCBI's E-Utilities API with provided term"""
+        search_url = f"{self.base_ncbi_uri}esearch.fcgi?db=gene&term={term}{self.api_key}&retmode=json"
+        try:
+            search_response = urllib.request.urlopen(search_url).read()
+        except Exception:
+            raise NCBIUnexpectedResultException
+        return json.loads(search_response)
+
+    def _is_valid_search_result(self, search_result):
+        """Checks that a search result contains only one UID as a result"""
         if (
-            "esearchresult" not in search_result
-            or "idlist" not in search_result["esearchresult"]
-            or len(search_result["esearchresult"]["idlist"]) < 1
-            or int(search_result["esearchresult"]["count"]) != 1
+            "esearchresult" in search_result
+            and "idlist" in search_result["esearchresult"]
+            and len(search_result["esearchresult"]["idlist"]) == 1
         ):
-            logging.error(f"Unexpected NCBI search result, got {search_result}")
-            return None
-        return int(search_result["esearchresult"]["idlist"][0])
+            try:
+                int(search_result["esearchresult"]["idlist"][0])
+                return True
+            except Exception:
```

Comment:
can you add an exception type here?

Reply:
you could also catch (and log) `IndexError` and `KeyError` and remove the if statement on liness 92-95